samples/RefreshFlow/AngularApp/src/core/auth-token/auth-token.service.ts

```diff
@@ -0,0 +1,133 @@
+                private loggedInActions: LoggedInActions,
+                private authTokenActions: AuthTokenActions,
+                private authReadActions: AuthReadyActions,
+                private profileActions: ProfileActions
+    ) { }
+
+
+    getTokens(data: LoginModel | RefreshGrant, grantType: string): Observable<void> {
+        // data can be any since it can either be a refresh tokens or login details
+        // The request for tokens must be x-www-form-urlencoded
+        let headers = new Headers({ 'Content-Type': 'application/x-www-form-urlencoded'});
+        let options = new RequestOptions({ headers: headers });
+
+        Object.assign(data, {
+            grant_type: grantType,
+            // offline_access is required for a refresh token
+            scope: ['openid offline_access']
+        });
+
+        return this.http.post('http://localhost:5000/connect/token', this.encodeObjectToParams(data) , options)
+            .map( res => res.json())
+            .map( (tokens: AuthTokenModel) => {
+                this.store.dispatch(this.authTokenActions.load(tokens));
+                this.store.dispatch(this.loggedInActions.loggedIn());
+
+                let profile = this.jwtHelper.decodeToken(tokens.id_token) as ProfileModel;
+                this.store.dispatch(this.profileActions.Load(profile));
+
+                this.storage.setItem('auth-tokens', tokens);
+            })
+            .do( _ => this.store.dispatch(this.authReadActions.ready()))
+          //  .catch( error => this.httpExceptions.handleTokenBadRequest(error));
+
+    }
+
+    deleteTokens() {
+        this.storage.removeItem('auth-tokens');
+        this.store.dispatch(this.authTokenActions.delete());
+    }
+
+    unsubscribeRefresh() {
+        if (this.refreshSubscription$) {
+            this.refreshSubscription$.unsubscribe();
+        }
+    }
+
+    refreshTokens(): Observable<Response> {
+        return this.store.map( state => state.auth.authTokens.refresh_token)
+            .first()
+            .flatMap( refreshToken => {
+                return this.getTokens(
+                    { refresh_token: refreshToken } as RefreshGrant, 'refresh_token')
+                    .catch( error => Observable.throw('Session Expired'));
+            });
+    }
+
+    startupTokenRefresh() {
+        return this.storage.getItem('auth-tokens')
+            .flatMap( (tokens: AuthTokenModel) => {
+                // check if the token is even if localStorage, if it isn't tell them it's not and return
+                if (!tokens) {
+                    this.store.dispatch(this.authReadActions.ready());
+                    return Observable.throw('No token in Storage');
+                }
+                // parse the token into a model and throw it into the store
+                this.store.dispatch(this.authTokenActions.load(tokens));
+
+                if (!this.jwtHelper.isTokenExpired(tokens.id_token)) {
```

Comment:
Hum, this doesn't seem the right thing to do, as identity tokens usually expire before access tokens. To determine whether an access token is still valid, consider extracting the `expires_in` property returning in the token response. It's the only reliable way to know the access token expiration date.